---
title: "Inh correlation plots across 3 species"
output:
  html_document:
    df_print: paged
  pdf_document:
    keep_tex: yes
keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(Seurat)
library(cowplot)
library(dplyr)
library(qlcMatrix)
library(Hmisc)
```

load mouse data

```{r}
setwd('~/postdoc2/Gedankenpapers/CNevomanuscript/code/snRNAseq/chicken/')
load(file = '../mouse/data/dataforplotting.RData')

mouse<-VZ

Idents(mouse)<-'final.clusters2'
mouse<-RenameIdents(mouse,'Inh_1'='Inh1','Inh_2'='Inh2.1','Inh_3'='Inh2.2','Inh_4'='Inh2.3','Inh_5'='Inh3')
mouse$final.clusters3<-Idents(mouse)


mouse.data<-mouse@assays$RNA@counts
```

revert gene names to  ENSMUSG genes 
```{r}
#load conversion table
conversiontable=read.csv("../mouse/EnseblID_to_GeneSymbol.txt")
conversiontable.df<-as.data.frame(conversiontable)

conversiontable.notrepeated<-conversiontable[!(duplicated(conversiontable[,2])|duplicated(conversiontable[,2], fromLast=TRUE)),]
rownames(conversiontable.notrepeated)<-conversiontable.notrepeated[,1]


counts_translated<-mouse.data[rownames(mouse.data) %in% conversiontable.notrepeated[,2],]


rownames(counts_translated)<-as.character(conversiontable.notrepeated$Gene.stable.ID[match(rownames(counts_translated),conversiontable.notrepeated$Gene.name)])


counts_nottranslated<-mouse.data[!rownames(mouse.data) %in% conversiontable.notrepeated[,2],]

mouse.data2<-rbind(counts_translated,counts_nottranslated)

```







load chicken data
```{r}
load('data/inh_clean.RData')
all<-inh

chicken.data<-all@assays$RNA@counts
chicken<-all

Idents(chicken)<-'merged.res.2.renamed'
chicken<-RenameIdents(chicken,'Inh1'='Inh1.1','Inh2'='Inh1.2','Inh3'='Inh1.3','Inh4'='Inh2','Inh5'='Inh3')
chicken$final.clusters3<-Idents(chicken)

rm(inh)
```


revert gene names to  ENSGALG genes 
```{r}
#load conversion table
conversiontable=read.csv("data/EnsembleID99_to_GeneSymbol.txt")
conversiontable.df<-as.data.frame(conversiontable)

conversiontable.notrepeated<-conversiontable[!(duplicated(conversiontable[,2])|duplicated(conversiontable[,2], fromLast=TRUE)),]
rownames(conversiontable.notrepeated)<-conversiontable.notrepeated[,1]


counts_translated<-chicken.data[rownames(chicken.data) %in% conversiontable.notrepeated[,2],]


rownames(counts_translated)<-as.character(conversiontable.notrepeated$Gene.stable.ID[match(rownames(counts_translated),conversiontable.notrepeated$Gene.name)])


counts_nottranslated<-chicken.data[!rownames(chicken.data) %in% conversiontable.notrepeated[,2],]

chicken.data2<-rbind(counts_translated,counts_nottranslated)

```





load human data

```{r}
load('../human/data/dataforplotting_3donors.RData')
DefaultAssay(CN.all.integrated)<-'RNA'
all<-CN.all.integrated
human.data<-all@assays$RNA@counts
human<-all


Idents(human)<-'merged.res.2.renamed'
human$final.clusters3<-Idents(human)
```


revert gene names to  ENSG genes 
```{r}
#load conversion table
conversiontable=read.csv("../human/data/EnsemblID_to_GeneSymbol.txt")
conversiontable.df<-as.data.frame(conversiontable)

conversiontable.notrepeated<-conversiontable[!(duplicated(conversiontable[,2])|duplicated(conversiontable[,2], fromLast=TRUE)),]
rownames(conversiontable.notrepeated)<-conversiontable.notrepeated[,1]


counts_translated<-human.data[rownames(human.data) %in% conversiontable.notrepeated[,2],]


rownames(counts_translated)<-as.character(conversiontable.notrepeated$Gene.stable.ID[match(rownames(counts_translated),conversiontable.notrepeated$Gene.name)])


counts_nottranslated<-human.data[!rownames(human.data) %in% conversiontable.notrepeated[,2],]

human.data2<-rbind(counts_translated,counts_nottranslated)

```












read in ortholog relationships obtained from ensemble biomart

```{r}
orth.table<-read.csv(file='data/chicken_human_mouse_orthologs.txt',header=T)
#throw out anything that isnt onetoone
orth.table<-orth.table[orth.table[,"Human.homology.type"]=="ortholog_one2one" & orth.table[,"Mouse.homology.type"]=="ortholog_one2one",]
#drop extra columns
orth.table.clean<-unique(orth.table[,c("Gene.stable.ID",'Human.gene.stable.ID','Human.gene.name','Mouse.gene.stable.ID','Mouse.gene.name')])
```


throw out what isn't in a 1:1 ortholog
```{r}
chicken.data.filtered<-chicken.data2[rownames(chicken.data2) %in% as.character(orth.table.clean$Gene.stable.ID),]
rownames(chicken.data.filtered)<-as.character(orth.table.clean$Mouse.gene.stable.ID[match(rownames(chicken.data.filtered),orth.table.clean$Gene.stable.ID)])

human.data.filtered<-human.data2[rownames(human.data2) %in% as.character(orth.table.clean$Human.gene.stable.ID),]
rownames(human.data.filtered)<-as.character(orth.table.clean$Mouse.gene.stable.ID[match(rownames(human.data.filtered),orth.table.clean$Human.gene.stable.ID)])

mouse.data.filtered<-mouse.data2[rownames(mouse.data2) %in% as.character(orth.table.clean$Mouse.gene.stable.ID),]
rownames(mouse.data.filtered)<-as.character(orth.table.clean$Mouse.gene.stable.ID[match(rownames(mouse.data.filtered),orth.table.clean$Mouse.gene.stable.ID)])


chicken.data.filtered2<-chicken.data.filtered[rownames(chicken.data.filtered) %in% rownames(human.data.filtered) & rownames(chicken.data.filtered) %in% rownames(mouse.data.filtered),]
human.data.filtered2<-human.data.filtered[rownames(human.data.filtered) %in% rownames(chicken.data.filtered2) & rownames(human.data.filtered) %in% rownames(mouse.data.filtered),]

mouse.data.filtered2<-mouse.data.filtered[rownames(mouse.data.filtered) %in% rownames(chicken.data.filtered2) & rownames(mouse.data.filtered) %in% rownames(human.data.filtered2),]
```

now convert them all back to mouse names for simplicity.

replace known ENSMUSG genes with their names, keep the remainder
```{r}
#load conversion table
conversiontable=read.csv("../mouse/EnseblID_to_GeneSymbol.txt")
conversiontable.df<-as.data.frame(conversiontable)

conversiontable.notrepeated<-conversiontable[!(duplicated(conversiontable[,2])|duplicated(conversiontable[,2], fromLast=TRUE)),]
rownames(conversiontable.notrepeated)<-conversiontable.notrepeated[,1]


counts_translated<-chicken.data.filtered2[rownames(chicken.data.filtered2) %in% rownames(conversiontable.notrepeated),]
rownames(counts_translated)<-as.character(conversiontable.notrepeated$Gene.name[match(rownames(counts_translated),conversiontable.notrepeated$Gene.stable.ID)])
counts_nottranslated<-chicken.data.filtered2[!rownames(chicken.data.filtered2) %in% rownames(conversiontable.notrepeated),]
chicken.data.filtered3<-rbind(counts_translated,counts_nottranslated)


counts_translated<-human.data.filtered2[rownames(human.data.filtered2) %in% rownames(conversiontable.notrepeated),]
rownames(counts_translated)<-as.character(conversiontable.notrepeated$Gene.name[match(rownames(counts_translated),conversiontable.notrepeated$Gene.stable.ID)])
counts_nottranslated<-human.data.filtered2[!rownames(human.data.filtered2) %in% rownames(conversiontable.notrepeated),]
human.data.filtered3<-rbind(counts_translated,counts_nottranslated)

counts_translated<-mouse.data.filtered2[rownames(mouse.data.filtered2) %in% rownames(conversiontable.notrepeated),]
rownames(counts_translated)<-as.character(conversiontable.notrepeated$Gene.name[match(rownames(counts_translated),conversiontable.notrepeated$Gene.stable.ID)])
counts_nottranslated<-mouse.data.filtered2[!rownames(mouse.data.filtered2) %in% rownames(conversiontable.notrepeated),]
mouse.data.filtered3<-rbind(counts_translated,counts_nottranslated)

```

Set up seurat objects.

```{r}
human2<-CreateSeuratObject(human.data.filtered3)
human2@meta.data<-human@meta.data
human2$species<-'human'
Idents(human2)<-'final.clusters3'

chicken2<-CreateSeuratObject(chicken.data.filtered3)
chicken2@meta.data<-chicken@meta.data
chicken2$species<-'chicken'
Idents(chicken2)<-'final.clusters3'

mouse2<-CreateSeuratObject(mouse.data.filtered3)
mouse2@meta.data<-mouse@meta.data
mouse2$species<-'mouse'
Idents(mouse2)<-'final.clusters3'



human2<- NormalizeData(object = human2, verbose = FALSE,scale.factor = 1e6)
chicken2<- NormalizeData(object = chicken2, verbose = FALSE,scale.factor = 1e6)
mouse2<- NormalizeData(object = mouse2, verbose = FALSE,scale.factor = 1e6)


#chickenDEGs<-FindAllMarkers(chicken2,test.use = 'wilcox')
#humanDEGs<-FindAllMarkers(human2,test.use = 'wilcox')
#mouseDEGs<-FindAllMarkers(mouse2,test.use = 'wilcox')

#save(chickenDEGs,file='chickenDEGs_3species_inh.RData')
#save(humanDEGs,file='humanDEGs_3species_inh_3donors.RData')
#save(mouseDEGs,file='mouseDEGs_3species_inh.RData')



load('data/chickenDEGs_3species_inh.RData')
load('data/humanDEGs_3species_inh_3donors.RData')
load('data/mouseDEGs_3species_inh.RData')
```


filter DEGs
```{r}
humanDEGs.filtered<-humanDEGs[humanDEGs$p_val_adj<0.01 & humanDEGs$avg_logFC>log(2),]
chickenDEGs.filtered<-chickenDEGs[chickenDEGs$p_val_adj<0.01 & chickenDEGs$avg_logFC>log(2),]
mouseDEGs.filtered<-mouseDEGs[mouseDEGs$p_val_adj<0.01 & mouseDEGs$avg_logFC>log(2),]


human2@assays$RNA@var.features<-unique(humanDEGs.filtered$gene)
chicken2@assays$RNA@var.features<-unique(chickenDEGs.filtered$gene)
mouse2@assays$RNA@var.features<-unique(mouseDEGs.filtered$gene)


```


Calculate cluster averages.
```{r}
interestinggenes<-intersect(unique(humanDEGs.filtered$gene),unique(chickenDEGs.filtered$gene))
interestinggenes<-intersect(interestinggenes,unique(mouseDEGs.filtered$gene))
humanaverage<-AverageExpression(human2,features=interestinggenes,verbose=F)
chickenaverage<-AverageExpression(chicken2,features=interestinggenes,verbose = F)
mouseaverage<-AverageExpression(mouse2,features=interestinggenes,verbose = F)
```

normalize
```{r}
humanaverage<-humanaverage[[1]]
chickenaverage<-chickenaverage[[1]]
mouseaverage<-mouseaverage[[1]]

human.norm<-humanaverage/rowMeans(humanaverage)
chicken.norm<-chickenaverage/rowMeans(chickenaverage)
mouse.norm<-mouseaverage/rowMeans(mouseaverage)

colnames(human.norm)<-paste0('H_',colnames(human.norm))
colnames(chicken.norm)<-paste0('C_',colnames(chicken.norm))
colnames(mouse.norm)<-paste0('M_',colnames(mouse.norm))
```




calculate pvalues by shuffling gene expression values across cell types.

```{r}
repeats=10000
cuts=20000

p<-replicate(repeats,{
  ctest<-t(apply(chicken.norm,1,sample))
  colnames(ctest)<-colnames(chicken.norm)
  htest<-t(apply(human.norm,1,sample))
  colnames(htest)<-colnames(human.norm)
  mtest<-t(apply(mouse.norm,1,sample))
  colnames(mtest)<-colnames(mouse.norm)
  res<-rcorr(as.matrix(cbind(ctest,mtest,htest)),type="spearman")
  x<-res[[1]]
#  x<-as.vector(x[grep('chicken',rownames(x)),grep('mouse',colnames(x))])
  })

h<-hist(as.vector(p[p!=1]),breaks=seq(-1,1,2/cuts),freq=F,)

h2<-h$counts/sum(h$counts)
c<-cumsum(h2)
plot(c)

library(Hmisc)
res<-rcorr(as.matrix(cbind(chicken.norm,mouse.norm,human.norm)),type="spearman")
x<-res[[1]]
pval<-c[round((x+1)*cuts/2)]
pval2<-matrix(as.vector(pval),nrow=dim(x)[1],ncol=dim(x)[2],byrow = F)
rownames(pval2)<-rownames(x)
colnames(pval2)<-colnames(x)
```

calculate multiscale bootstrapped dendrogram
```{r,fig.width=15,fig.height=6}
library(heatmap3)
alpha=0.05

sig<-pval2<alpha | pval2> 1-alpha





library(ComplexHeatmap)
library(dendsort)
library(dendextend)
library(circlize)
library(pvclust)
result <- pvclust(as.matrix(cbind(chicken.norm,mouse.norm,human.norm)), nboot=10000,method.hclust = "complete",method.dist = function(z){as.dist(1-cor(z,use="pa",method="spearman"))})

```
plot dendrogram
```{r,fig.width=5,fig.height=2.5}

speciespal=c('gold','maroon','grey')
species=as.factor(c("C","C","C","C","C",'M','M',"M","M","M",'H','H',"H","H","H"))
levels(species)<-c(1,2,3)

VZpalchicken[5]<-'dimgrey'
VZpal[5]<-'dimgrey'

CLpal <- c(VZpalchicken,VZpal,VZpalhuman)                           
bars=data.frame(CL=CLpal,animal=speciespal[(species)]);




dend <- (as.dendrogram(result))
dend %>% pvclust_show_signif_gradient(result,signif_type = 'au',
                                      signif_col_fun = colorRampPalette(c('white','black'))) %>%
  set("nodes_pch", 19) %>%
  set("branches_lwd", 1) %>%
   plot()
colored_bars(colors = bars, dend = dend, sort_by_labels_order = TRUE)

result %>% text


```

plot correlation matrix.

```{r,fig.width=6,fig.height=4.75}
library(ComplexHeatmap)
library(dendsort)
library(dendextend)
library(circlize)

col1=colorRamp2(c(-1, 0, 1), c("navy", "white", "firebrick3"),space="sRGB")

dend1=dend
dend2=dend

mat<-x
#colnames(mat)<-colnames(mouseaverage)
#rownames(mat)<-colnames(chickenaverage)


#column annotation
#ha_bottom<- HeatmapAnnotation(
#  cl=colnames(mat),
#  col=list(cl=setNames(c(VZpal,RLpal[9]),colnames(mat)[c(2:5,1,6)])),
#  col=list(cl=setNames(c(VZpal),colnames(mat))),

#    simple_anno_size=unit(2,"mm")
#)

#row annotation
#ha_right<- rowAnnotation(
#  cl=rownames(mat),
#  col=list(cl=setNames(VZpalchicken,rownames(mat))),
#  simple_anno_size=unit(2,"mm")
#)

Heatmap(mat,
        name='correlation',
        col=col1,
#        bottom_annotation = ha_bottom,
#        right_annotation = ha_right,
#        clustering_distance_rows  = function(x) as.dist(1 - cor(t(x), use = "pa")),
#        clustering_distance_columns  = function(x) as.dist(1 - cor(t(x), use = "pa")),
#        row_dend_reorder =FALSE, column_dend_reorder = F,
        cluster_rows = (dend1),cluster_columns = (dend2),
        row_dend_width = unit(25, "mm"),
        column_dend_height = unit(25,"mm"),
        row_split=3,column_split = 3,
        border=T,
        cell_fun = function(j, i, x, y, width, height, fill) {
            if(sig[i,j]>0)
            grid.points(x, y,pch=16,size = unit(2, "mm"))
          }
)
```

```{r}

```

